daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala

```diff
@@ -455,6 +497,674 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
+               ubind x: M:T <- fetch @M:T cId
+               in Test:exercise_by_key exercisingParty (Test:someParty sig) Test:noCid 0 (M:Either:Left @Int64 @Int64 0)""",
+            Array(SParty(charlie), SContractId(cId), SParty(alice)),
+            Set(alice, charlie),
+            getContract = getContract,
+            getKey = getKey,
+          )
+
+          inside(res) { case Success(Left(SErrorDamlException(IE.FailedAuthorization(_, _)))) =>
+            msgs shouldBe Seq(
+              "starts test",
+              "maintainers",
+              "queries key",
+              "choice controllers",
+              "choice observers",
+            )
+          }
+        }
+      }
+
+      "a local contract" - {
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of successful exercise_by_key of a local contract
+        "success" in {
+          val (res, msgs) = evalUpdateApp(
+            pkgs,
+            e"""\(sig: Party) (obs : Party) (exercisingParty : Party) ->
+             ubind cId: ContractId M:T <- create @M:T M:T { signatory = sig, observer = obs, precondition = True, key = M:toKey sig, nested = M:buildNested 0 } in 
+             Test:exercise_by_key exercisingParty (Test:someParty sig) Test:noCid 0 (M:Either:Left @Int64 @Int64 0)
+             """,
+            Array(SParty(alice), SParty(bob), SParty(alice)),
+            Set(alice),
+          )
+          inside(res) { case Success(Right(_)) =>
+            msgs shouldBe Seq(
+              "starts test",
+              "maintainers",
+              "choice controllers",
+              "choice observers",
+              "choice body",
+              "ends test",
+            )
+          }
+        }
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of exercise_by_key of an inactive local contract
+        "inactive contract" in {
+          val (res, msgs) = evalUpdateApp(
+            pkgs,
+            e"""\(sig : Party) (obs : Party) (exercisingParty : Party) ->
+             ubind 
+               cId: ContractId M:T <- create @M:T M:T { signatory = sig, observer = obs, precondition = True, key = M:toKey sig, nested = M:buildNested 0 } ;
+               x: Unit <- exercise @M:T Archive cId sig
+             in 
+               Test:exercise_by_key exercisingParty (Test:someParty sig) Test:noCid 0 (M:Either:Left @Int64 @Int64 0)
+             """,
+            Array(SParty(alice), SParty(bob), SParty(alice)),
+            Set(alice),
+          )
+          inside(res) { case Success(Left(SErrorDamlException(IE.ContractKeyNotFound(gKey)))) =>
+            gKey.templateId shouldBe T
+            msgs shouldBe Seq(
+              "starts test",
+              "maintainers",
+            )
+          }
+        }
+
+        // TEST_EVIDENCE: Semantics: Evaluation order of exercise_by_key of a cached global contract with failure authorization
+        "authorization failures" in {
```

Comment:
should we add visibility failures here as well?

Reply:
It was forgotten.

Reply:
added in fef1db52a0c28175482905a99701932c08ecb02f